File: tests/extract.go
package tests

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

var (
	reLeader      = regexp.MustCompile(`<[Ll]eader>`)
	reCtrlPrefix  = regexp.MustCompile(`<[Cc]-`)
	reCtrlChord   = regexp.MustCompile(`^<c-([A-Za-z0-9]+)>`)
	reLuaMap      = regexp.MustCompile(`map\s*\(\s*"[nv]"\s*,\s*"([^"]+)"`)
	reLuaKeymap   = regexp.MustCompile(`keymap\.set\s*\(\s*"[nv]"\s*,\s*"([^"]+)"`)
	reLuaKeymapM  = regexp.MustCompile(`keymap\.set\s*\(\s*\{[^}]+\}\s*,\s*"([^"]+)"`)
	reVimLeader   = regexp.MustCompile(`<[Ll]eader>([^\s:]*)`)
	reVimCtrl     = regexp.MustCompile(`(<[Cc]-[A-Za-z0-9]+>)`)
	reVimRedo     = regexp.MustCompile(`(^|\s)nmap\s+U\s+`)
	reVimComment  = regexp.MustCompile(`^\s*"`)
	reVimCoc      = regexp.MustCompile(`nmap\s+<silent>\s*(\w+)`)
)

func normalize(lhs string) string {
	lhs = strings.TrimSpace(lhs)
	lhs = reLeader.ReplaceAllString(lhs, "<leader>")
	lhs = reCtrlPrefix.ReplaceAllString(lhs, "<c-")
	if m := reCtrlChord.FindStringSubmatch(lhs); m != nil {
		return "<c-" + strings.ToLower(m[1]) + ">"
	}
	return lhs
}

func fromLua(text string, keys map[string]struct{}) {
	for _, re := range []*regexp.Regexp{reLuaMap, reLuaKeymap, reLuaKeymapM} {
		for _, m := range re.FindAllStringSubmatch(text, -1) {
			keys[normalize(m[1])] = struct{}{}
		}
	}
}

func fromVim(text string, keys map[string]struct{}) {
	for _, line := range strings.Split(text, "\n") {
		if reVimComment.MatchString(line) {
			continue
		}
		if m := reVimLeader.FindStringSubmatch(line); m != nil {
			keys[normalize("<leader>"+m[1])] = struct{}{}
		}
		if m := reVimCtrl.FindStringSubmatch(line); m != nil {
			keys[normalize(m[1])] = struct{}{}
		}
		if reVimRedo.MatchString(line) {
			keys["U"] = struct{}{}
		}
		if strings.Contains(line, "<Plug>(coc") {
			if m := reVimCoc.FindStringSubmatch(line); m != nil {
				keys[m[1]] = struct{}{}
			}
		}
	}
}

func detect(root string) string {
	if fileExists(filepath.Join(root, "lua", "mapping", "basis.lua")) {
		return "current"
	}
	if fileExists(filepath.Join(root, ".vim", "Mapping.vim")) || fileExists(filepath.Join(root, "init.vim")) {
		return "vim"
	}
	return ""
}

func collect(root string) (string, map[string]struct{}, error) {
	flavor := detect(root)
	keys := map[string]struct{}{}
	switch flavor {
	case "current":
		err := filepath.WalkDir(filepath.Join(root, "lua", "mapping"), func(path string, d os.DirEntry, err error) error {
			if err != nil || d.IsDir() || !strings.HasSuffix(path, ".lua") {
				return err
			}
			b, err := os.ReadFile(path)
			if err != nil {
				return err
			}
			fromLua(string(b), keys)
			return nil
		})
		return flavor, keys, err
	case "vim":
		for _, rel := range []string{filepath.Join(".vim", "Mapping.vim"), "init.vim"} {
			p := filepath.Join(root, rel)
			if !fileExists(p) {
				continue
			}
			b, err := os.ReadFile(p)
			if err != nil {
				return flavor, keys, err
			}
			fromVim(string(b), keys)
		}
		return flavor, keys, nil
	default:
		return flavor, keys, nil
	}
}

func present(keys map[string]struct{}, key string) bool {
	switch key {
	case "u":
		_, a := keys["U"]
		_, b := keys["u"]
		return a || b
	case "j":
		_, a := keys["J"]
		_, b := keys["j"]
		return a || b
	case "k":
		_, a := keys["K"]
		_, b := keys["k"]
		return a || b
	}
	_, ok := keys[key]
	return ok
}

func fileExists(path string) bool {
	st, err := os.Stat(path)
	return err == nil && !st.IsDir()
}

File: tests/extract.lua
-- Pull map lhs values out of Lua or VimScript files. No Neovim required.

local extract = {}

local function exists(path)
  local f = io.open(path, "r")
  if f then
    f:close()
    return true
  end
  return false
end

local function read(path)
  local f = io.open(path, "r")
  if not f then
    return ""
  end
  local data = f:read("*a")
  f:close()
  return data or ""
end

local function list_files(dir)
  local files = {}
  local pipe = io.popen('find "' .. dir .. '" -type f 2>/dev/null')
  if not pipe then
    return files
  end
  for line in pipe:lines() do
    files[#files + 1] = line
  end
  pipe:close()
  return files
end

function extract.normalize(lhs)
  lhs = lhs:gsub("^%s+", ""):gsub("%s+$", "")
  lhs = lhs:gsub("<[Ll]eader>", "<leader>")
  lhs = lhs:gsub("<[Cc]%-", "<c-")
  lhs = lhs:gsub("^<c%-([%w]+)>", function(inner)
    return "<c-" .. inner:lower() .. ">"
  end)
  return lhs
end

local function add(set, lhs)
  if lhs and lhs ~= "" then
    set[extract.normalize(lhs)] = true
  end
end

function extract.from_lua(text, set)
  for lhs in text:gmatch('map%s*%(%s*"[nv]"%s*,%s*"([^"]+)"') do
    add(set, lhs)
  end
  for lhs in text:gmatch('keymap%.set%s*%(%s*"[nv]"%s*,%s*"([^"]+)"') do
    add(set, lhs)
  end
  for lhs in text:gmatch('keymap%.set%s*%(%s*{[^}]+}%s*,%s*"([^"]+)"') do
    add(set, lhs)
  end
end

function extract.from_vim(text, set)
  for line in (text .. "\n"):gmatch("(.-)\n") do
    if not line:match("^%s*\"") then
      local suffix = line:match("<[Ll]eader>([^%s:]*)")
      if suffix then
        add(set, "<leader>" .. suffix)
      end
      local chord = line:match("(<%s*[Cc]%-[%w]+>)")
      if chord then
        add(set, chord)
      end
      if line:match("%f[%w]nmap%s+U%s+") then
        add(set, "U")
      end
      local coc = line:match("<Plug>%(coc%-%w+%)") and line:match("%f[%w]nmap%s+<silent>%s*(%w+)")
      if coc then
        add(set, coc)
      end
    end
  end
end

function extract.detect(root)
  if exists(root .. "/lua/mapping/basis.lua") then
    return "current"
  end
  if exists(root .. "/.vim/Mapping.vim") then
    return "vim"
  end
  if exists(root .. "/init.vim") then
    return "vim"
  end
  return nil
end

function extract.collect(root)
  local flavor = extract.detect(root)
  local set = {}
  if flavor == "current" then
    for _, path in ipairs(list_files(root .. "/lua/mapping")) do
      if path:match("%.lua$") then
        extract.from_lua(read(path), set)
      end
    end
  elseif flavor == "vim" then
    if exists(root .. "/.vim/Mapping.vim") then
      extract.from_vim(read(root .. "/.vim/Mapping.vim"), set)
    end
    if exists(root .. "/init.vim") then
      extract.from_vim(read(root .. "/init.vim"), set)
    end
  end
  return flavor, set
end

return extract
